tools/proofers/text_operations.js

```diff
@@ -0,0 +1,187 @@
+        // * Illustration markup may appear w/o a title, so remove the ': '.
+        // * Formatting markup is redundant w/o any content, so don't produce it.
+        if (selection === '') {
+            if (tagOpen === '[Illustration: ') {
+                tagOpen = '[Illustration';
+            } else if ((tagOpen[0] === '<') && (tagOpen.length > 1)) {
+                // do not tag empty strings but insert a single '<'
+                tagOpen = '';
+                tagClose = '';
+            }
+        }
+
+        // Handle footnote label substitution
+        if (tagOpen === '[Footnote #: ') {
+            // Split the selected text on the first space in the string.
+            // If the first part is a label use it in the opening tag in
+            // place of '#', otherwise remove the ' #' from the opening tag.
+            var label = '';
+            var i = selection.indexOf(' ');
+            if (i !== -1) {
+                var first = selection.substr(0, i);
+
+                // A string is a footnote label if it's a letter A-Z, or an integer > 0
+                if ((/^[A-Za-z]$|^[1-9]\d*$/).test(first)) {
+                    label = ' ' + first;
+                    selection = selection.substr(i + 1);
+                }
+            }
+            tagOpen = tagOpen.replace(' #', label);
+
+            // If there's no selection, remove the label entirely.
+            if (selection === '') {
+                tagOpen = tagOpen.replace(': ', '');
+            }
+        } else if (tagOpen === '[** ' && selection) {
+            tagOpen = selection + tagOpen;
+        }
+
+        var subst = tagOpen + selection + tagClose;
+        txtArea.value = txtArea.value.substring(0, startPos) + subst + txtArea.value.substring(endPos);
+        var curPos = startPos + subst.length;
+        txtArea.setSelectionRange(retainSelection ? startPos : curPos, curPos);
+        focusTrigger();
+    }
+
+    function lcCommon(str) {
+        var words = str.split(' ');
+        var commonLcWords = ':At:Under:Near:Upon:By:Of:In:On:For' + // prepositions
+                        ':Is:Was:Are' +    // 'small' verbs
+                        ':But:And:Or' +    // conjunctions
+                        ':A:An:The' +      // articles
+                        ':Am:Pm:Bc:Ad' +   // small caps abbreviations
+                        ':De:Van:La:Le:';  // LOTE
+
+        // Start at i=1 to avoid changing the first word (leave it Titlecased).
+        // E.g. if str is "A Winter's Tale", we don't want to lowercase the "A".
+        var i;
+        for(i = 1; i < words.length; i += 1) {
+            // If the word appears in the :-delimited list above, it should be lower case
+            if (commonLcWords.indexOf(':' + words[i] + ':') !== -1) {
+                words[i] = words[i].toLowerCase();
+            }
+        }
+
+        return words.join(' ');
+    }
+
+    function titleCase(str) {
+        str = str.toLowerCase();
+        var newStr = '';
```

Comment:
I know some of this is old, but should we modernize with let, const, for of etc.?

Reply:
Yes.